xds/src/main/java/io/grpc/xds/XdsLrsClient.java

```diff
@@ -0,0 +1,347 @@
+import io.grpc.internal.GrpcUtil;
+import io.grpc.stub.StreamObserver;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Client of XDS load reporting service. Methods in this class are expected to be called in
+ * the same synchronized context that {@link XdsLoadBalancer#helper#getSynchronizationContext}
+ * returns.
+ */
+@NotThreadSafe
+final class XdsLrsClient implements XdsLoadStatsManager {
+
+  @VisibleForTesting
+  static final String TRAFFICDIRECTOR_HOSTNAME_FIELD
+      = "com.googleapis.trafficdirector.grpc_hostname";
+  private final String serviceName;
+  private final ManagedChannel channel;
+  private final SynchronizationContext syncContext;
+  private final ScheduledExecutorService timerService;
+  private final Supplier<Stopwatch> stopwatchSupplier;
+  private final Stopwatch retryStopwatch;
+  private final ChannelLogger logger;
+  private final BackoffPolicy.Provider backoffPolicyProvider;
+  private final XdsLoadReportStore loadReportStore;
+  private boolean started;
+
+  @Nullable
+  private BackoffPolicy lrsRpcRetryPolicy;
+  @Nullable
+  private ScheduledHandle lrsRpcRetryTimer;
+
+  @Nullable
+  private LrsStream lrsStream;
+
+  XdsLrsClient(ManagedChannel channel,
+      Helper helper,
+      BackoffPolicy.Provider backoffPolicyProvider) {
+    this(channel, helper, GrpcUtil.STOPWATCH_SUPPLIER, backoffPolicyProvider,
+        new XdsLoadReportStore(checkNotNull(helper, "helper").getAuthority()));
+  }
+
+  @VisibleForTesting
+  XdsLrsClient(ManagedChannel channel,
+      Helper helper,
+      Supplier<Stopwatch> stopwatchSupplier,
+      BackoffPolicy.Provider backoffPolicyProvider,
+      XdsLoadReportStore loadReportStore) {
+    this.channel = checkNotNull(channel, "channel");
+    this.serviceName = checkNotNull(helper.getAuthority(), "serviceName");
+    this.syncContext = checkNotNull(helper.getSynchronizationContext(), "syncContext");
+    this.stopwatchSupplier = checkNotNull(stopwatchSupplier, "stopwatchSupplier");
+    this.retryStopwatch = stopwatchSupplier.get();
+    this.logger = checkNotNull(helper.getChannelLogger(), "logger");
+    this.timerService = checkNotNull(helper.getScheduledExecutorService(), "timeService");
+    this.backoffPolicyProvider = checkNotNull(backoffPolicyProvider, "backoffPolicyProvider");
+    this.loadReportStore = checkNotNull(loadReportStore, "loadReportStore");
+    started = false;
+  }
+
+  @Override
+  public void startLoadReporting() {
+    checkState(!started, "load reporting has already started");
+    started = true;
+    startLrsRpc();
+  }
+
+  @Override
+  public void stopLoadReporting() {
+    if (lrsRpcRetryTimer != null) {
+      lrsRpcRetryTimer.cancel();
+    }
+    if (lrsStream != null) {
+      lrsStream.close(null);
+    }
+    started = false;
```

Comment:
Will `startLoadReporting()` ever be called after `stopLoadReporting()`? I see `stopLoadReporting()` more like `shutdown()` that is only called when XdsLoadBalancer is shutting down. Then you just don't reset `started`.

p.s. Allowing that case is problematic anyway, because `startLrsRpc()` may throw because `lrsStream` may be non-null.

Reply:
Agree. I don't think we have the use case to restart the load reporting after shutting it down. I guess the LRS process will stay on as long as the channel to the remote balancer exists. Stopping load reporting should only happen upon shutting down the `XdsLoadBalancer`.

Reply:
ACK.